src/components/FileList.jsx

```diff
@@ -1,14 +1,78 @@
-import React from 'react'
+import React, { PureComponent } from 'react'
 import { translate } from '../lib/I18n'
 
-import File from '../components/File'
+import InfiniteLoader from 'react-virtualized/dist/commonjs/InfiniteLoader'
+import AutoSizer from 'react-virtualized/dist/commonjs/AutoSizer'
+import List from 'react-virtualized/dist/commonjs/List'
+import File, { FilePlaceholder } from '../components/File'
 
-const FileList = ({
-  t, f, displayedFolder, files, selected, selectionModeActive, onFileEdit, onFolderOpen, onFileOpen, onFileToggle, showActionMenu
-}) => (
-  <div>
-    {files.map((file, idx) => (
+const STATUS_LOADING = 1
+const STATUS_LOADED = 2
+
+const FETCH_LIMIT = 50
+
+class FileList extends PureComponent {
+  constructor (props) {
+    super(props)
+    this.requestedFiles = {}
+    for (var i = 0; i < props.files.length; i++) this.requestedFiles[i] = STATUS_LOADED
+  }
+
+  render () {
+    const { fileCount, files, selected } = this.props
+    return (
+      <InfiniteLoader
+        isRowLoaded={this.isRowLoaded}
+        loadMoreRows={this.loadMoreRows}
+        rowCount={fileCount}
+        minimumBatchSize={FETCH_LIMIT}
+      >
+        {({ onRowsRendered, registerChild }) => (
+          <AutoSizer>
+            {({ width, height }) => (
+              <List
+                ref={registerChild}
+                height={height}
+                rowCount={fileCount}
+                onRowsRendered={onRowsRendered}
+                rowHeight={50}
+                rowRenderer={this.rowRenderer}
+                width={width}
+                selected={selected}
+                files={files}
+              />
+            )}
+          </AutoSizer>
+        )}
+      </InfiniteLoader>
+    )
+  }
+
+  isRowLoaded = ({ index }) => {
+    return !!this.requestedFiles[index]
```

Comment:
No entry in this map signifies that the row has never been loaded before. An entry (either LOADING or LOADED) can be treated as loaded as far as InfiniteLoader is concerned.